src/commonMain/kotlin/fr/acinq/lightning/blockchain/electrum/SwapInWallet.kt

```diff
@@ -0,0 +1,52 @@
+package fr.acinq.lightning.blockchain.electrum
+
+import fr.acinq.lightning.NodeParams
+import fr.acinq.lightning.blockchain.electrum.WalletState.Companion.index
+import fr.acinq.lightning.crypto.KeyManager
+import fr.acinq.lightning.utils.MDCLogger
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.filterNotNull
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.launch
+import org.kodein.log.LoggerFactory
+import org.kodein.log.newLogger
+
+class SwapInWallet(
+    chain: NodeParams.Chain,
+    swapInKeys: KeyManager.SwapInOnChainKeys,
+    electrum: IElectrumClient,
+    addressGenerationWindow: Int,
+    scope: CoroutineScope,
+    loggerFactory: LoggerFactory
+) {
+    private val logger = loggerFactory.newLogger(this::class)
+
+    val wallet = ElectrumMiniWallet(chain.chainHash, electrum, scope, logger)
+
+    val legacySwapInAddress: String = swapInKeys.legacySwapInProtocol.address(chain)
+        .also { wallet.addAddress(it) }
+    val swapInAddressFlow = MutableStateFlow<Pair<String, Int>?>(null)
+        .also { wallet.addAddressGenerator(generator = { index -> swapInKeys.getSwapInProtocol(index).address(chain) }, window = addressGenerationWindow) }
+
+    init {
+        scope.launch {
+            // address rotation
+            wallet.walletStateFlow.map { it ->
+                // take the first unused address with the lowest index
+                it.addresses
+                    .filter { it.value.unspent.isEmpty() }
+                    .mapNotNull { (key, value) -> value.meta.index?.let { key to it } }
+                    .minByOrNull { it.second }
+            }
+                .filterNotNull()
+                .distinctUntilChanged()
+                .collect { (address, index) ->
+                    logger.info { "setting current swap-in address=$address index=$index" }
+                    swapInAddressFlow.emit(address to index)
```

Comment:
If the wallet has already used N addresses, we do not want to emit all those addresses sequentially at startup until we detect the first unused address.

Current implementation guarantees that we will not emit addresses that have already been used, because an address will not be added to `WalletState` before we receive the first `ScriptHashSubscriptionResponse` from Electrum.